table.cpp

```diff
@@ -211,96 +224,143 @@ void table_t::start()
+                         // clang-format on
+                         "    COLLATE \"C\"") %
+                     name %
+                     (table_space ? "TABLESPACE " + table_space.get() : ""))
+                        .str());
+            }
+            pgsql_exec_simple(sql_conn, PGRES_COMMAND_OK,
+                              (fmt("DROP TABLE %1%") % name).str());
             pgsql_exec_simple(
                 sql_conn, PGRES_COMMAND_OK,
-                (fmt("CREATE TABLE %1%_tmp %2% AS\n"
-                     "  SELECT * FROM %1%\n"
-                     "    ORDER BY ST_GeoHash(way,10)\n"
-                     "    COLLATE \"C\"") %
-                 name % (table_space ? "TABLESPACE " + table_space.get() : ""))
-                    .str());
+                (fmt("ALTER TABLE %1%_tmp RENAME TO %1%") % name).str());
+            fprintf(stderr, "Copying %s to cluster by geometry finished\n",
+                    name.c_str());
         } else {
-            /* osm2pgsql's transformation from 4326 to another projection could make a geometry invalid,
-               and these need to be filtered. Also, a transformation is needed for geohashing. */
-            pgsql_exec_simple(
-                sql_conn, PGRES_COMMAND_OK,
-                (fmt("CREATE TABLE %1%_tmp %2% AS\n"
-                     "  SELECT * FROM %1%\n"
-                     "    WHERE ST_IsValid(way)\n"
-                     // clang-format off
-                     "    ORDER BY ST_GeoHash(ST_Transform(ST_Envelope(way),4326),10)\n"
-                     // clang-format on
-                     "    COLLATE \"C\"") %
-                 name % (table_space ? "TABLESPACE " + table_space.get() : ""))
-                    .str());
-        }
-        pgsql_exec_simple(sql_conn, PGRES_COMMAND_OK, (fmt("DROP TABLE %1%") % name).str());
-        pgsql_exec_simple(sql_conn, PGRES_COMMAND_OK, (fmt("ALTER TABLE %1%_tmp RENAME TO %1%") % name).str());
-        fprintf(stderr, "Copying %s to cluster by geometry finished\n", name.c_str());
-        fprintf(stderr, "Creating geometry index on %s\n", name.c_str());
-
-        // Use fillfactor 100 for un-updatable imports
-        pgsql_exec_simple(sql_conn, PGRES_COMMAND_OK, (fmt("CREATE INDEX ON %1% USING GIST (way) %2% %3%") % name %
-            (slim && !drop_temp ? "" : "WITH (FILLFACTOR=100)") %
-            (table_space_index ? "TABLESPACE " + table_space_index.get() : "")).str());
-
-        /* slim mode needs this to be able to apply diffs */
-        if (slim && !drop_temp)
-        {
-            fprintf(stderr, "Creating osm_id index on %s\n", name.c_str());
-            pgsql_exec_simple(sql_conn, PGRES_COMMAND_OK, (fmt("CREATE INDEX ON %1% USING BTREE (osm_id) %2%") % name %
-                (table_space_index ? "TABLESPACE " + table_space_index.get() : "")).str());
+            // We haven't removed invalid geometries as we did not perform a COPY.
```

Comment:
But this can IMHO be forced on inital parameter checks the way that "--skip-(optimiz|index)ing" is only allowed when "--unlogged" is requested as well - after all, the user should know what he is doing by requesting it.